source/extensions/filters/common/ext_proc/grpc_client.h

```diff
@@ -0,0 +1,300 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/config/core/v3/grpc_service.pb.h"
+#include "envoy/grpc/async_client_manager.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stream_info/stream_info.h"
+
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/http/sidestream_watermark.h"
+#include "source/common/runtime/runtime_features.h"
+#include "source/extensions/filters/common/ext_proc/client_base.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace ExternalProcessing {
+
+/**
+ * Callbacks used by the processor stream.
+ */
+template <typename ResponseType> class ProcessorCallbacks : public RequestCallbacks<ResponseType> {
+public:
+  virtual ~ProcessorCallbacks() = default;
+  virtual void onReceiveMessage(std::unique_ptr<ResponseType>&& response) PURE;
+  virtual void onGrpcError(Grpc::Status::GrpcStatus error, const std::string& message) PURE;
+  virtual void onGrpcClose() PURE;
+  virtual void logStreamInfo() PURE;
+};
+
+/**
+ * A stream to an external processing server. This uses a bidirectional gRPC stream to send
+ * requests and receive responses.
+ */
+template <typename RequestType, typename ResponseType> class ProcessorStream : public StreamBase {
+public:
+  ~ProcessorStream() override = default;
+  virtual void send(RequestType&& request, bool end_stream) PURE;
+  // Idempotent close. Return true if it actually closed.
+  virtual bool close() PURE;
+  virtual bool halfCloseAndDeleteOnRemoteClose() PURE;
+  virtual const StreamInfo::StreamInfo& streamInfo() const PURE;
+  virtual StreamInfo::StreamInfo& streamInfo() PURE;
+  virtual void notifyFilterDestroy() PURE;
+};
+
+template <typename RequestType, typename ResponseType>
+using ProcessorStreamPtr = std::unique_ptr<ProcessorStream<RequestType, ResponseType>>;
+
+/**
+ * Implementation of ProcessorStream that uses a gRPC service.
+ */
+template <typename RequestType, typename ResponseType>
+class ProcessorStreamImpl : public ProcessorStream<RequestType, ResponseType>,
+                            public Grpc::AsyncStreamCallbacks<ResponseType>,
+                            public Logger::Loggable<Logger::Id::ext_proc> {
+public:
+  // Factory method: create and return `ProcessorStreamPtr`; return nullptr on failure.
+  static ProcessorStreamPtr<RequestType, ResponseType>
+  create(Grpc::AsyncClient<RequestType, ResponseType>&& client,
+         ProcessorCallbacks<ResponseType>& callbacks, Http::AsyncClient::StreamOptions& options,
+         Http::StreamFilterSidestreamWatermarkCallbacks& sidestream_watermark_callbacks,
+         const std::string& service_method);
+
+  void send(RequestType&& request, bool end_stream) override;
+  // Close the stream. This is idempotent and will return true if we
+  // actually closed it.
+  bool close() override;
+  bool halfCloseAndDeleteOnRemoteClose() override;
+
+  void notifyFilterDestroy() override {
```

Comment:
Do you want to move the definition outside of the class declaration, same as the other member functions?

Reply:
This is a small class, and I tend to keep here to keep it grouped.